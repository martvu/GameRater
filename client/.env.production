VITE_SERVER_URI=http://it2810-48.idi.ntnu.no:4000/graphql