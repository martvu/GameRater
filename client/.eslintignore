dist
.eslintrc.js
.prettierrc.js
src/gql/