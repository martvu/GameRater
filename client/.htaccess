<IfModule mod_rewrite.c>
  RewriteEngine On
  RewriteBase /project2/
  RewriteRule ^index\.html$ - [L]
  RewriteCond %{REQUEST_FILENAME} !-f
  RewriteCond %{REQUEST_FILENAME} !-d
  RewriteRule . /project2/index.html [L]
</IfModule>
