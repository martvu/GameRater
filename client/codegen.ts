import type { CodegenConfig } from '@graphql-codegen/cli';

const config: CodegenConfig = {
  schema: '../server/src/schema.graphql',
  documents: ['src/**/*.{ts,tsx}'],
  generates: {
    './src/gql/': {
      preset: 'client',
      plugins: [],
      overwrite: true,
      presetConfig: {
        gqlTagName: 'gql',
        fragmentMasking: false,
      },
    },
  },
};
export default config;
