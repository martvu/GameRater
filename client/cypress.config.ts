import { defineConfig } from 'cypress';
import { connect, disconnect } from './cypress/support/db';

export default defineConfig({
  e2e: {
    baseUrl: 'http://localhost:5173/project2',
    setupNodeEvents(on) {
      // This task cleans the reviews and users collections in the database
      on('task', {
        async clearDB() {
          const db = await connect();
          const reviews = db.collection('reviews');
          await reviews.deleteMany({});
          await reviews.dropIndexes();
          const users = db.collection('users');
          await users.deleteMany({});
          await users.dropIndexes();
          await disconnect();
          return null;
        },
      });
    },
  },
});
