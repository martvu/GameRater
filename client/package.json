{
  "name": "prosjekt-2",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "scripts": {
    "dev": "npm run generate && vite --host",
    "build": "tsc && vite build",
    "lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
    "preview": "vite preview",
    "test": "vitest",
    "coverage": "vitest run --coverage",
    "generate": "graphql-codegen",
    "cypress:open": "cypress open"
  },
  "dependencies": {
    "@apollo/client": "^3.8.6",
    "@hookform/resolvers": "^3.3.2",
    "@radix-ui/react-avatar": "^1.0.4",
    "@radix-ui/react-checkbox": "^1.0.4",
    "@radix-ui/react-dialog": "^1.0.5",
    "@radix-ui/react-dropdown-menu": "^2.0.6",
    "@radix-ui/react-label": "^2.0.2",
    "@radix-ui/react-scroll-area": "^1.0.5",
    "@radix-ui/react-select": "^2.0.0",
    "@radix-ui/react-slot": "^1.0.2",
    "@radix-ui/react-toast": "^1.1.5",
    "@radix-ui/react-tooltip": "^1.0.7",
    "class-variance-authority": "^0.7.0",
    "clsx": "^2.0.0",
    "graphql": "^16.8.1",
    "lucide-react": "^0.284.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-hook-form": "^7.47.0",
    "react-infinite-scroll-component": "^6.1.0",
    "react-router-dom": "^6.16.0",
    "recoil": "^0.7.7",
    "tailwind-merge": "^1.14.0",
    "tailwindcss-animate": "^1.0.7",
    "zod": "^3.22.4"
  },
  "devDependencies": {
    "@graphql-codegen/cli": "^5.0.0",
    "@graphql-codegen/client-preset": "^4.1.0",
    "@testing-library/jest-dom": "^6.1.3",
    "@testing-library/react": "^14.0.0",
    "@testing-library/user-event": "^14.5.1",
    "@types/jest": "^29.5.5",
    "@types/node": "^20.8.6",
    "@types/react": "^18.2.15",
    "@types/react-dom": "^18.2.7",
    "@typescript-eslint/eslint-plugin": "^6.0.0",
    "@typescript-eslint/parser": "^6.0.0",
    "@vitejs/plugin-react-swc": "^3.3.2",
    "@vitest/coverage-v8": "^0.34.6",
    "autoprefixer": "^10.4.16",
    "cypress": "^13.5.1",
    "eslint": "^8.45.0",
    "eslint-config-prettier": "^9.0.0",
    "eslint-plugin-prettier": "^5.0.1",
    "eslint-plugin-react-hooks": "^4.6.0",
    "eslint-plugin-react-refresh": "^0.4.3",
    "jsdom": "^22.1.0",
    "postcss": "^8.4.31",
    "prettier": "^3.0.3",
    "prettier-plugin-tailwindcss": "^0.5.5",
    "tailwindcss": "^3.3.3",
    "typescript": "^5.2.2",
    "vite": "^4.4.5",
    "vitest": "^0.34.6"
  }
}
