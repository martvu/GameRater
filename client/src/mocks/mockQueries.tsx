import {
  ADD_FAVORITES,
  CREATE_REVIEW,
  REMOVE_FAVORITES,
  SIGN_IN_OR_CREATE_USER,
} from '@/lib/mutations';
import {
  GET_GAMES,
  GET_GAME,
  GET_FILTERS,
  GET_GAME_PLATFORMS,
  GET_SEARCH_SUGGESTIONS,
} from '@/lib/queries';

const getGamesMock = {
  request: {
    query: GET_GAMES,
    variables: {
      limit: 24,
      offset: 0,
      sortBy: { field: 'first_release_date', order: 'desc' },
      platforms: [],
      genres: [],
      query: undefined,
      userId: '',
      showFavorites: false,
      showReviewedGames: false,
    },
  },
  result: {
    data: {
      search: {
        count: 1,
        games: [
          {
            _id: '1',
            aggregated_rating: 85,
            first_release_date: 1620000000,
            summary: 'Example game summary',
            cover_image_id: 'cover1',
            name: 'Example Game',
            user_rating: 4.5,
          },
        ],
        filters: {
          genres: [1, 2], // Example genre IDs
          platforms: [3, 4], // Example platform IDs
        },
      },
    },
  },
};

const getGamesMockedKeyword = {
  request: {
    query: GET_GAMES,
    variables: {
      limit: 24,
      offset: 0,
      sortBy: { field: 'first_release_date', order: 'desc' },
      platforms: [],
      genres: [],
      query: 'mockedKeyword',
      userId: '',
      showFavorites: false,
      showReviewedGames: false,
    },
  },
  result: {
    data: {
      search: {
        count: 1,
        games: [
          {
            _id: '1',
            aggregated_rating: 85,
            first_release_date: 1620000000,
            summary: 'Example game summary',
            cover_image_id: 'cover1',
            name: 'Example Game',
            user_rating: 4.5,
          },
        ],
        filters: {
          genres: [1, 2], // Example genre IDs
          platforms: [3, 4], // Example platform IDs
        },
      },
    },
  },
};

const getGamesSignedInMock = {
  request: {
    query: GET_GAMES,
    variables: {
      limit: 24,
      offset: 0,
      sortBy: { field: 'first_release_date', order: 'desc' },
      platforms: [],
      genres: [],
      query: 'mockedKeyword',
      userId: '1',
      showFavorites: false,
      showReviewedGames: false,
    },
  },
  result: {
    data: {
      search: {
        count: 1,
        games: [
          {
            _id: '1',
            aggregated_rating: 85,
            first_release_date: 1620000000,
            summary: 'Example game summary',
            cover_image_id: 'cover1',
            name: 'Example Game',
            platforms: [1, 2],
            genres: [3, 4],
          },
          // Add more game objects as needed
        ],
        filters: {
          genres: [1, 2], // Example genre IDs
          platforms: [3, 4], // Example platform IDs
        },
      },
    },
  },
};

const getFiltersMock = {
  request: {
    query: GET_FILTERS,
    variables: {},
  },
  result: {
    data: {
      getGenres: [
        { id: 1, name: 'Genre 1' },
        { id: 2, name: 'Genre 2' },
      ],
      getPlatforms: [
        { id: 1, name: 'Platform 1' },
        { id: 3, name: 'Platform 3' },
        { id: 4, name: 'PC (Microsoft Windows)' },
        { id: 2, name: 'Nintendo Switch' },
        { id: 5, name: 'PlayStation 5' },
        { id: 6, name: 'Xbox Series X|S' },
      ],
    },
  },
};

const getGamePlatformsMock = {
  request: {
    query: GET_GAME_PLATFORMS,
    variables: { id: '1' },
  },
  result: {
    data: {
      getGame: {
        name: 'Example Game',
        platforms: [
          {
            id: 3,
            name: 'Platform 3',
          },
          {
            id: 4,
            name: 'PC (Microsoft Windows)',
          },
        ],
      },
    },
  },
};

const signInMock = {
  request: {
    query: SIGN_IN_OR_CREATE_USER,
    variables: {
      userInput: {
        username: 'testuser',
      },
    },
  },
  result: {
    data: {
      signInOrCreateUser: {
        _id: '1',
        username: 'testuser',
        favorites: [],
        reviews: [],
      },
    },
  },
};

const getGameMock = {
  request: {
    query: GET_GAME,
    variables: {
      id: '1',
      limit: 5,
      offset: 0,
      username: '',
    },
  },
  result: {
    data: {
      getGame: {
        _id: '1',
        name: 'Example Game',
        summary: 'This is an example game summary.',
        imageId: 'image1',
        first_release_date: 1620000000,
        aggregatedRating: 85,
        user_rating: 4.5,
        platforms: [{ name: 'PC' }],
        genres: [{ name: 'Adventure' }],
        reviews: {
          count: 2,
          reviews: [
            {
              _id: 'review1',
              user: 'user1',
              title: 'Great Game',
              content: 'This is a great game because...',
              rating: 5,
              platform: 'PC',
              gameID: '1',
            },
            {
              _id: 'review2',
              user: 'user2',
              title: 'Good Game',
              content: 'This is a good game because...',
              rating: 4,
              platform: 'PC',
              gameID: '1',
            },
          ],
          userHasReviewed: false,
        },
      },
    },
  },
};

const getGameMock2 = {
  request: {
    query: GET_GAME,
    variables: {
      id: '1',
      limit: 5,
      offset: 0,
      username: 'testUser',
    },
  },
  result: {
    data: {
      getGame: {
        _id: '1',
        name: 'Example Game',
        summary: 'This is an example game summary.',
        imageId: 'image1',
        first_release_date: 1620000000,
        aggregatedRating: 85,
        user_rating: 4.5,
        platforms: [{ name: 'PC' }],
        genres: [{ name: 'Adventure' }],
        reviews: {
          count: 2,
          reviews: [
            {
              _id: 'review1',
              user: 'user1',
              title: 'Great Game',
              content: 'This is a great game because...',
              rating: 5,
              platform: 'PC',
              gameID: '1',
            },
            {
              _id: 'review2',
              user: 'user2',
              title: 'Good Game',
              content: 'This is a good game because...',
              rating: 4,
              platform: 'PC',
              gameID: '1',
            },
          ],
          userHasReviewed: false,
        },
      },
    },
  },
};
const addFavoritesMock = {
  request: {
    query: ADD_FAVORITES,
    variables: { username: 'testUser', gameID: '1' },
  },
  result: {
    data: {
      addFavorites: {
        _id: '1',
        username: 'testUser',
        favorites: [{ _id: '1' }],
      },
    },
  },
};

const removeFavoritesMock = {
  request: {
    query: REMOVE_FAVORITES,
    variables: { username: 'testUser', gameID: '1' },
  },
  result: {
    data: {
      removeFavorites: {
        _id: '1',
        username: 'testUser',
        favorites: [],
      },
    },
  },
};

const createReviewMock = {
  request: {
    query: CREATE_REVIEW,
    variables: {
      reviewInput: {
        title: 'Great Game',
        content: 'This is a great game because...',
        rating: 5,
        platform: 'PC (Microsoft Windows)',
        user: 'testUser',
        gameID: '1',
      },
    },
  },
  result: {
    data: {
      createReview: {
        _id: 'review1',
        user: 'testUser',
        title: 'Great Game',
        content: 'This is a great game because...',
        rating: 5,
        platform: 'PC (Microsoft Windows)',
        gameID: '1',
      },
    },
  },
};

const searchbarSuggestionsMock = {
  request: {
    query: GET_SEARCH_SUGGESTIONS,
    variables: {
      query: 'example',
    },
  },
  result: {
    data: {
      getSearchSuggestions: [
        {
          _id: '1',
          name: 'Example Game',
          cover_image_id: 'cover1',
        },
        {
          _id: '2',
          name: 'Example Game 2',
          cover_image_id: 'cover2',
        },
      ],
    },
  },
};

const allMocks = [
  createReviewMock,
  getGamePlatformsMock,
  getGamesMockedKeyword,
  addFavoritesMock,
  removeFavoritesMock,
  getGamesMock,
  getFiltersMock,
  signInMock,
  getGamesSignedInMock,
  getGameMock,
  getGameMock2,
  searchbarSuggestionsMock,
  // Add more mocks as needed for other queries
];

export {
  allMocks,
  getGamesMock,
  getGamesSignedInMock,
  getFiltersMock,
  signInMock,
};
