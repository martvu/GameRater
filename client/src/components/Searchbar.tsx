import { Label } from '@/components/ui/label.tsx';
import { Input } from '@/components/ui/input.tsx';
import { Button } from '@/components/ui/button.tsx';
import { Search, XCircle } from 'lucide-react';
import {
  FocusEvent,
  FormEvent,
  KeyboardEvent,
  useEffect,
  useRef,
  useState,
} from 'react';
import { Link, useLocation, useNavigate, useParams } from 'react-router-dom';
import { searchQueryState } from '@/state/atoms.ts';
import { useRecoilState } from 'recoil';
import { useLazyQuery } from '@apollo/client';
import { GET_SEARCH_SUGGESTIONS } from '@/lib/queries.tsx';
import { cn } from '@/lib/utils.ts';
import ProgressiveImage from '@/components/ProgressiveImage.tsx';
import imageNotFound from '@/assets/img-fallback.svg';

type SearchbarProps = {
  showFullWidthSearch?: boolean;
};

/**
 * Searchbar component
 * @param {boolean} [showFullWidthSearch] - Whether to show the full width search bar or not
 */
const Searchbar = ({ showFullWidthSearch }: SearchbarProps) => {
  const navigate = useNavigate();
  const location = useLocation();
  const { keyword } = useParams<{ keyword?: string }>();
  const [searchQuery, setSearchQuery] = useRecoilState(searchQueryState);
  const [isFocused, setIsFocused] = useState(false);
  const [getGames, { called, loading, data }] = useLazyQuery(
    GET_SEARCH_SUGGESTIONS,
    {
      fetchPolicy: 'no-cache', // Use 'cache-and-network' or 'network-only' if you want to always fetch the latest results
    }
  );
  const dropdownRef = useRef(null);

  // Debounce the search query (Wait for 300ms after the user stops typing)
  useEffect(() => {
    const timerId = setTimeout(() => {
      if (searchQuery.trim()) {
        getGames({ variables: { query: searchQuery } });
      }
    }, 300);

    return () => clearTimeout(timerId);
  }, [searchQuery, getGames]);

  const handleBlur = (event: FocusEvent) => {
    // Delay hiding dropdown to allow for click events to be processed
    setTimeout(() => {
      const relatedTarget = event.relatedTarget as Node | null;
      const dropdown = dropdownRef.current as HTMLUListElement | null;

      if (dropdown && !dropdown.contains(relatedTarget)) {
        setIsFocused(false);
      }
    }, 100);
  };
  const handleFocus = () => {
    setIsFocused(true);
  };

  const handleInputKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'ArrowDown') {
      event.preventDefault();
      const dropdown = dropdownRef.current as HTMLUListElement | null;
      const firstItem = dropdown?.querySelector('li');
      (firstItem as HTMLElement)?.focus();
    } else if (event.key === 'Escape') {
      event.currentTarget.blur();
      setIsFocused(false);
    }
  };
  const handleLinkClick = () => {
    setIsFocused(false); // Hide the dropdown
    ((document.activeElement as HTMLElement) || null)?.blur(); // Remove focus from the active element
  };
  const handleListItemKeyDown = (event: KeyboardEvent<HTMLLIElement>) => {
    // Get the parent list item of the focused element
    const listItem = event.currentTarget;

    switch (event.key) {
      case 'ArrowDown':
      case 'ArrowUp': {
        event.preventDefault(); // Stop the page from scrolling

        const nextListItem =
          event.key === 'ArrowDown'
            ? (listItem.nextElementSibling as HTMLElement | null)
            : (listItem.previousElementSibling as HTMLElement | null);

        nextListItem?.focus();
        break;
      }
      case 'Enter': {
        // Simulate a click on the Link component when Enter is pressed
        const link = listItem.querySelector('a') as HTMLAnchorElement | null;
        if (link) {
          link.click(); // Perform the navigation
          setIsFocused(false); // Hide the dropdown
          ((document.activeElement as HTMLElement) || null)?.blur(); // Remove focus from the active element
        }
        break;
      }
    }
  };
  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    setIsFocused(false);
    ((document.activeElement as HTMLElement) || null)?.blur();
    navigate(
      searchQuery === '' || searchQuery === null
        ? '/'
        : `/search/${encodeURIComponent(searchQuery)}`
    ); // Navigate to the new URL
  };

  return (
    <>
      <form onSubmit={handleSubmit} className="flex">
        <Label htmlFor="search" className="sr-only" />
        <div
          className={`relative ${
            showFullWidthSearch ? 'w-full' : 'md:w-[350px] lg:w-[400px]'
          }`}
        >
          <Input
            id="search"
            data-testid="search-input"
            type="text"
            className="w-full rounded-xl bg-muted pr-24 shadow-inner md:w-full lg:w-full"
            placeholder={keyword || 'Search'}
            value={searchQuery}
            onChange={e => setSearchQuery(e.target.value)}
            onFocus={handleFocus}
            onBlur={handleBlur}
            onKeyDown={handleInputKeyDown}
            autoComplete="off"
          />
          {searchQuery && (
            <Button
              aria-label="Empty search input"
              data-testid="empty-search-input"
              variant="ghost"
              size="round"
              type="button"
              onClick={() => {
                setSearchQuery('');
                if (!location.pathname.startsWith('/game/')) navigate('/');
              }}
              className="absolute bottom-0 right-10 top-0 flex items-center justify-center"
            >
              <XCircle size={18} />
            </Button>
          )}
          <Button
            aria-label="Search"
            variant="search"
            type="submit"
            className="absolute right-0 top-0 rounded-l-none rounded-r-xl px-3 active:opacity-90"
          >
            <Search />
          </Button>
        </div>
      </form>
      {isFocused && searchQuery && (
        <ul
          aria-live="polite"
          aria-label="search suggestions"
          ref={dropdownRef}
          className={cn(
            'absolute top-full z-10 rounded-md border border-accent bg-background shadow-lg',
            showFullWidthSearch
              ? 'w-full'
              : 'w-[220px] md:w-[320px] lg:w-[420px]'
          )}
        >
          {loading && <li className="p-2">Loading...</li>}
          {!loading &&
            data &&
            data.getSearchSuggestions.map(game => (
              <li
                key={game._id}
                className="p-2 hover:bg-accent"
                tabIndex={0}
                data-testid="search-item"
                role="listitem"
                onKeyDown={handleListItemKeyDown}
              >
                <Link
                  aria-label={`Link to ${game.name} detail page`}
                  to={`/game/${game._id}`}
                  onClick={handleLinkClick}
                >
                  <div className="flex gap-4">
                    <ProgressiveImage
                      fullSrc={`https://images.igdb.com/igdb/image/upload/t_cover_small/${game.cover_image_id}.jpg`}
                      placeholderSrc={imageNotFound}
                      alt={game.name as string}
                      className="h-14 w-10 rounded-full object-cover"
                    />
                    <p className="text-left text-sm font-semibold">
                      {game.name}
                    </p>
                  </div>
                </Link>
              </li>
            ))}
          {called &&
            !loading &&
            data &&
            data.getSearchSuggestions.length === 0 && (
              <li className="p-2 text-center">No results found.</li>
            )}
        </ul>
      )}
    </>
  );
};

export default Searchbar;
