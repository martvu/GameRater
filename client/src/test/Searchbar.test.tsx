import { describe, it, expect } from 'vitest';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { RecoilRoot } from 'recoil';
import { allMocks } from '@/mocks/mockQueries';
import userEvent from '@testing-library/user-event';
import Searchbar from '@/components/Searchbar.tsx';
import { MemoryRouter } from 'react-router-dom';

interface SearchbarOptions {
  showFullWidthSearch?: boolean;
}

function renderSearchbar({
  showFullWidthSearch = false,
}: SearchbarOptions = {}) {
  return render(
    <MockedProvider mocks={allMocks} addTypename={false}>
      <RecoilRoot>
        <MemoryRouter initialEntries={['/project2']}>
          <Searchbar showFullWidthSearch={showFullWidthSearch} />
        </MemoryRouter>
      </RecoilRoot>
    </MockedProvider>
  );
}

describe('Searchbar Component', () => {
  it('matches the snapshot', () => {
    const { asFragment } = renderSearchbar();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly', async () => {
    renderSearchbar();
    expect(screen.getByTestId('search-input')).toBeInTheDocument();
  });

  it('display dropdown when input has input', async () => {
    renderSearchbar();
    await userEvent.type(screen.getByTestId('search-input'), 'example');
    await waitFor(() => {
      expect(screen.getByText('Example Game')).toBeInTheDocument();
      expect(screen.getByText('Example Game 2')).toBeInTheDocument();
    });
  });

  it('allows navigation through the dropdown using keyboard', async () => {
    renderSearchbar();

    // Simulate typing to trigger the dropdown
    await userEvent.type(screen.getByTestId('search-input'), 'example');

    // Wait for the dropdown to appear and display items
    await waitFor(() => {
      expect(screen.getByText('Example Game')).toBeInTheDocument();
    });
    const input = screen.getByTestId('search-input');
    const dropdownItems = screen.getAllByRole('listitem'); // This assumes each dropdown item has a role='listitem'

    // Ensure the input is initially focused
    expect(input).toHaveFocus();

    // Simulate pressing ArrowDown to move focus to the first dropdown item
    await userEvent.keyboard('{ArrowDown}');

    // Check if the first dropdown item is focused
    expect(dropdownItems[0]).toHaveFocus();

    // Simulate pressing ArrowDown again to move focus to the second dropdown item
    await userEvent.keyboard('{ArrowDown}');

    // Check if the second dropdown item is focused
    expect(dropdownItems[1]).toHaveFocus();

    // Simulate pressing ArrowUp to move focus back to the first dropdown item
    await userEvent.keyboard('{ArrowUp}');

    // Check if the first dropdown item is focused again
    expect(dropdownItems[0]).toHaveFocus();

    fireEvent.click(screen.getAllByTestId('search-item')[0]);
    // Simulate pressing Enter to navigate to the first dropdown item
    await userEvent.keyboard('{Enter}');

    // Check if the dropdown is hidden
    expect(screen.queryByText('Example Game')).not.toBeInTheDocument();
  });
});
