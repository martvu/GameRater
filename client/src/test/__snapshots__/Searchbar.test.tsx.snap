// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Searchbar Component > matches the snapshot 1`] = `
<DocumentFragment>
  <form
    class="flex"
  >
    <label
      class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 sr-only"
      for="search"
    />
    <div
      class="relative w-[200px] md:w-[300px] lg:w-[400px]"
    >
      <input
        autocomplete="off"
        class="flex h-10 border border-input px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 w-full rounded-xl bg-muted pr-24 shadow-inner md:w-full lg:w-full"
        data-testid="search-input"
        id="search"
        placeholder="Search"
        type="text"
        value=""
      />
      <button
        aria-label="Search"
        class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-transparent text-secondary-foreground hover:text-muted-foreground transition duration-300 ease-in-out h-10 py-2 absolute right-0 top-0 rounded-l-none rounded-r-xl px-3 active:opacity-90"
        type="submit"
      >
        <svg
          class="lucide lucide-search"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <path
            d="m21 21-4.3-4.3"
          />
        </svg>
      </button>
    </div>
  </form>
</DocumentFragment>
`;
